### PR TITLE
Run gather bootstrap if recommended after OCP install fail

### PR DESCRIPTION
**Pull request by clacroix12**
Signed-off-by: Coady LaCroix <clacroix@redhat.com>

### CODE REVIEW

**Comment by clacroix12** [comment]
You can see a successful test of this [here](https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/4330/). In this case the removal of the bootstrap node was commented out and a temporary test case that executed the `gather_bootstrap` function was used to verify. You can see the result of the command [here](http://magna002.ceph.redhat.com/ocsci-jenkins/openshift-clusters/clacroix/clacroix_20200205T182458/logs/gather_bootstrap/). This test was done in vmware. AWS was tested locally prior to testing in vmware and was successful as well.

I am labeling `Needs Deployment Testing` as technically the branch changed since I ran since I had to remove temporary workarounds to force the test. I am anticipating that the default workflow (when we do not encounter a failure in which we run `gather_bootstrap` to remain unaffected).

**Comment by clacroix12** [comment]
> PR validation [Job FAILED](https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-trigger-test-pr/1024//testReport) (installation failed, tests not executed).

The actual scenario in which gather bootstrap is supposed to execute was why this PR test failed, and the command was never run. I see the proper string in the raised CommandFailed exception, however we still didn't execute gather_bootstrap. I need to do some troubleshooting to figure out why it's not parsing that correctly.

File: ocs_ci/utility/bootstrap.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+import boto3
+from libcloud.compute.providers import get_driver
+from libcloud.compute.types import Provider
+
+from ocs_ci.framework import config
+from ocs_ci.ocs import constants
+from ocs_ci.ocs.exceptions import NodeNotFoundError, UnsupportedPlatformError
+from ocs_ci.utility import vsphere
+from ocs_ci.utility.utils import run_cmd, get_infra_id
```
Comment (petr-balogh) [inline ocs_ci/utility/bootstrap.py]: `from ocs_ci.utility.utils get_infra_id, import run_cmd` .  alpha order

File: ocs_ci/utility/bootstrap.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def gather_bootstrap():
+    """
+    Gather debugging data for a failing-to-bootstrap control plane.
+    Data is placed in the `gather_bootstrap` directory under the log directory.
+
```
Comment (petr-balogh) [inline ocs_ci/utility/bootstrap.py]: Add raises section please and document: NodeNotFoundError + CommandFailed I think is also very possible to see here.

File: ocs_ci/utility/bootstrap.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    run_cmd(cmd)
+
+
+def get_gather_bootstrap_node_data():
+    """
+    Retrieve node IPs required by the gather bootstrap command
+
+    Returns:
+        dict: Public IP of the bootstrap node and Private IPs of master nodes
+
```
Comment (petr-balogh) [inline ocs_ci/utility/bootstrap.py]: Add Raises section please

File: ocs_ci/utility/bootstrap.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        )
+
+
+def get_node_data_aws():
+    """
+    Retrieve bootstrap public IP and master node private IPs
+
+    Returns:
+        dict: bootstrap and master node IP data
+
```
Comment (petr-balogh) [inline ocs_ci/utility/bootstrap.py]: Guess what :)

File: ocs_ci/utility/bootstrap.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        node for node in driver.list_nodes()
+        if master_pattern in node.name
+    ]
+    master_ips = [master.private_ips[0] for master in master_nodes]
+    data['master_ips'] = master_ips
+    logger.debug(data)
+    return data
+
+
+def get_node_data_vsphere():
```
Comment (petr-balogh) [inline ocs_ci/utility/bootstrap.py]: Dostring with return and raises

File: ocs_ci/deployment/aws.py [hunk 6]
```diff
@@ -14,19 +13,18 @@
...
+from ocs_ci.ocs import constants, exceptions, ocp
 from ocs_ci.ocs.parallel import parallel
+from ocs_ci.ocs.resources import pod
+from ocs_ci.utility import templating
 from ocs_ci.utility.aws import AWS as AWSUtil
+from ocs_ci.utility.bootstrap import gather_bootstrap
 from ocs_ci.utility.retry import retry
 from ocs_ci.utility.utils import (
-    run_cmd, clone_repo, create_rhelpod, TimeoutSampler
+    run_cmd, clone_repo, create_rhelpod, TimeoutSampler, get_infra_id,
```
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/aws.py]: If this is something we want to hold PRs to I suggest we add it to our [coding guidelines](https://github.com/red-hat-storage/ocs-ci/blob/master/docs/coding_guidelines.md)

File: ocs_ci/deployment/aws.py [hunk 7]
```diff
@@ -224,12 +222,20 @@ def deploy(self, log_cli_level='DEBUG'):
...
+                    f"--dir {self.cluster_path} "
+                    f"--log-level {log_cli_level}",
+                    timeout=3600
+                )
+            except exceptions.CommandFailed as e:
+                gather_bootstrap_pattern = (
+                    "openshift-install gather bootstrap --help"
+                )
+                if gather_bootstrap_pattern in str(e):
+                    gather_bootstrap()
```
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/aws.py]: I definitely think we should log the original error, but I don't really like the idea of wrapping it in additional text. It just clutters up an already descriptive error message.

File: ocs_ci/utility/bootstrap.py (hunk 2 above)
Comment (clacroix12) [reply ocs_ci/utility/bootstrap.py]: Do we include `CommandFailed` in other docstrings where that is not explicitly raised, but rather can be raised by subsequently called functions? I think our docstring should only reflect what this function raises explicitly, no?

File: ocs_ci/deployment/aws.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/aws.py]: @petr-balogh I am now calling `gather_bootstrap` inside a try/except where we will except all Exceptions from `gather_bootstrap` and simply log them at error level. This way the original error will still end up being raise no matter what happens during `gather_bootstrap`.

File: ocs_ci/deployment/aws.py [hunk 8]
```diff
@@ -399,6 +406,11 @@ def deploy(self, log_cli_level='DEBUG'):
 
             if proc.returncode:
                 logger.error(stderr)
+                gather_bootstrap_pattern = (
+                    "openshift-install gather bootstrap --help"
```
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/aws.py]: Done

File: ocs_ci/deployment/vmware.py [hunk 9]
```diff
@@ -449,12 +450,21 @@ def deploy(self, log_cli_level='DEBUG'):
...
-                timeout=3600
-            )
+            try:
+                run_cmd(
+                    f"{self.installer} wait-for bootstrap-complete "
+                    f"--dir {self.cluster_path} "
+                    f"--log-level {log_cli_level}",
+                    timeout=3600
+                )
+            except CommandFailed as e:
```
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/vmware.py]: Done

File: ocs_ci/utility/bootstrap.py (hunk 1 above)
Comment (clacroix12) [reply ocs_ci/utility/bootstrap.py]: Done

File: ocs_ci/deployment/aws.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (clacroix12) [reply ocs_ci/deployment/aws.py]: Done

File: ocs_ci/utility/bootstrap.py (hunk 2 above)
Comment (clacroix12) [reply ocs_ci/utility/bootstrap.py]: Done

File: ocs_ci/utility/bootstrap.py (hunk 3 above)
Comment (clacroix12) [reply ocs_ci/utility/bootstrap.py]: Done

File: ocs_ci/utility/bootstrap.py (hunk 4 above)
Comment (clacroix12) [reply ocs_ci/utility/bootstrap.py]: Done

File: ocs_ci/utility/bootstrap.py (hunk 5 above)
Comment (clacroix12) [reply ocs_ci/utility/bootstrap.py]: Done

File: ocs_ci/utility/bootstrap.py (hunk 2 above)
Comment (petr-balogh) [reply ocs_ci/utility/bootstrap.py]: yep, it would be better to stick just with raised exception in the documented function

**Pull request merged**